magento/pwa-studio
# Add middleware to schedule sign-out

## Description
When a user signs in, we get a token that we store in both app state (Redux store's `user` slice) and local storage (`signin_token`, namespaced). This token has a limited lifetime before it becomes invalid (currently hardcoded to **one hour**), and an invalid token is removed from storage the first time it is accessed after expiration.

Unfortunately, the storage interface can't notify app state when the token expires or when it removes an expired token, since it doesn't have access to the `dispatch` function. This leads app state and storage to fall out of sync, such that components start to throw errors when app state tells them the user is signed in but the token has actually been invalidated.

This PR adds a middleware that schedules the app state's `signOut` workflow to run whenever a token is set (sign-in or re-auth). Just as the token expires, the app will sign the user out, clear the token and all user state, and refresh the page. In the future, we can remove the page refresh, but that will require some amount of component rework and regression.

## Related Issue
PWA-1105

## Acceptance

### Verification Stakeholders
- @dpatil-magento 
- @sirugh 

### Specification

### Verification Steps
1. Sign in to a user account.
2. Open developer tools and edit the `signin_token` in storage to have a `ttl` of `30000` (30 seconds).
3. Refresh your browser tab (to simulate having waited approximately one hour).
4. Wait 30 seconds.
5. Verify that you are automatically signed out and the page automatically refreshes.
6. Verify that there are no JS errors in the console.
7. Repeat this process on the checkout page.

## Screenshots / Screen Captures (if appropriate)

## Checklist
* I have added tests to cover my changes, if necessary.
* I have added translations for new strings, if necessary.
* I have updated the documentation accordingly, if necessary.
